test/model/src/include/model/core.h

```diff
@@ -0,0 +1,437 @@
+/*-
+ * Public Domain 2014-present MongoDB, Inc.
+ * Public Domain 2008-2014 WiredTiger, Inc.
+ *
+ * This is free and unencumbered software released into the public domain.
+ *
+ * Anyone is free to copy, modify, publish, use, compile, sell, or
+ * distribute this software, either in source code form or as a compiled
+ * binary, for any purpose, commercial or non-commercial, and by any
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef MODEL_CORE_H
+#define MODEL_CORE_H
+
+#include <deque>
+#include <limits>
+#include <map>
+#include <mutex>
+#include <optional>
+#include <string>
+
+/* Redefine important WiredTiger internal constants, if they are not already available. */
+
+/*
+ * WT_TS_MAX --
+ *     The maximum timestamp, typically used in reads where we would like to get the latest value.
+ */
+#ifndef WT_TS_MAX
```

Comment:
Could this be renamed `WT_TS_LATEST` or similar? The current name does not hint at this behavior.

Reply:
Done. The C WiredTiger code has `WT_TS_MAX`, which is why I introduced it. But I also added the convenience alias for MAX as you suggested, as I too got confused while working on this code.

Reply:
Have you considered using `constexpr` so that you can move these definitions into the `model` namespace?
The interplay between this an the implicit WT namespace is unclear to me.

Reply:
This one is somewhat of a judgment call. I chose going with the C way because `WT_TS_NONE` in the model is now exactly the same constant as the one defined in `src/include/txn.h`. I put an `#ifndef` around it in the case the caller has the constant already imported from WiredTiger.

I'm saying it's a judgment call, because you are absolutely right that if I was going to define a new constant, it would totally make sense to use a C++ `constexpr`. The question is whether it makes sense to have a different name. I guess I already crossed the line as I defined `timestamp_t` in the namespace instead of redefining `wt_timestamp_t` from `src/include/wt_internal.h` exactly—and also with the other naming... I can see that. I'm not yet sure what I think of it yet, but maybe there is a value in just completely divorcing the naming and just making use of `static_assert` to ensure numerical equality if need be?

Reply:
I ended up switching to `constexpr`, as we've been already departing from C in this model. I'm still planning to keep static asserts to ensure that the model constants and the WT constants are numerically equal to avoid surprises later when using the model.